integration/openshift: Add OpenShift e2e tests runner

wainersm opened the pull request:
This adds wrapper scripts and the new `openshift-e2e` make target
which are used to run the OpenShift e2e tests.

Fixes: #2860
Depends-on: github.com/kata-containers/runtime#2969

Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

bergwolf commented:
@wainersm Thanks a lot! I'm wondering how are we going to test openshift? Are you planning to add some Jenkins jobs to validate it?

wainersm commented:
> @wainersm Thanks a lot! I'm wondering how are we going to test openshift? Are you planning to add some Jenkins jobs to validate it?

Hi @bergwolf ! I'm working to get Kata Containers on-boarded into OpenShift CI.

The biggest chunk of code for the initial support was in https://github.com/kata-containers/tests/pull/2802 and I still need to get merged this https://github.com/kata-containers/runtime/pull/2891, then this  https://github.com/openshift/release/pull/9009 on OpenShift side. Once all those are merged we going to start seeing nightly CI jobs for Kata + OpenShift in https://prow.ci.openshift.org

I am sending this PR so that I can later replace the smoke test (which will be executed in CI) by the OpenShift e2e tests.

wainersm commented:
/test

wainersm commented:
/cc @c3d @fidencio

cpmeadors commented:
@wainersm Is there any way to run the ci tests against my own cluster? Currently reading the code to see if I can figure it out.

wainersm commented:
> @wainersm Is there any way to run the ci tests against my own cluster? Currently reading the code to see if I can figure it out.

Yes, it was designed to allow running in the user's own cluster.

You need to export KUBECONFIG which points to your OpenShift cluster, then you run `make openshift-e2e`.

Notice it assumes that your cluster has kata containers properly installed and it was configured with a kata  runtimeClass and the kata admission controller deployed. If you don't have your cluster configured, you can export `CI=true` and  it will make the scripts to configure the cluster (beware it won't rollback to its previous state afterwards).

You can run it in dry-run mode, i.e., openshift e2e tests won't be executed, so just export`OPENSHIFT_TESTS_DRY_RUN=true`

chavafg commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,103 @@
+	configure_cluster || die "Failed to configure the cluster"
+else
+	info "Check the kata admission controller is active"
+	is_kata_admission_controller_active || exit 1
```
use `die` instead of `exit 1`?

chavafg commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,156 @@
+add_kata_runtimeclass() {
+	local handler=${1:-kata}
+	info "Add the ${kata_runtimeclass} runtimeClass"
+        if is_runtimeclass_present; then
```
can you please check indentation?

chavafg commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,156 @@
+#
+configure_cluster() {
+	info "Configure the OpenShift cluster"
+	add_kata_runtimeclass
```
could this be `add_kata_runtimeclass || die ""Failed to create the runtimeClass` ? Or does this necessarily need to be a Warning message?

wainersm replied:
I made this way so that the function's caller should decide whether to exit the script or not in case of failure.

chavafg commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,77 @@
+# Define temporary files and write down the patterns. Don't delete the files
+# afterwards because they can be used to debug.
+#
+inclusion_file=${script_dir}/tests_inclusion_patterns
```
could you quote variables here and below?

chavafg reviewed (commented):
Thanks @wainersm, looks good, just small nits.

cpmeadors commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,156 @@
+		return 1
+	fi
+
+	# Create a hello world pod and check if it is using the runtimeClass.
+	#
+	curl -s https://raw.githubusercontent.com/openshift/origin/master/examples/hello-openshift/hello-pod.json |\
+		oc apply -f - &>/dev/null
+	local class_name=$(oc get -o jsonpath='{.spec.runtimeClassName}' \
+		pod/${pod})
+	oc delete pod/${pod} &>/dev/null
+	if [ "$class_name" != "$kata_runtimeclass" ]; then
+		warn "runtimeClass: expected ${kata_runtimeclass}, got ${class_name}"
+		return 1
+	fi
```
I am not seeing a runtimeClass specified in the hello-pod.json.  I don't see how it can ever succeed the check to see if the class_name is equate to kata_runtimeclass.  Am I missing something?

wainersm replied:
If the admission control is up and running correctly then regular pods (e.g. this hello-pod) will have their containers attached the Kata runtimeClass. In simple words, the admission controller will make almost all pods to use Kata runtime even if the pod descriptor doesn't say so.

wainersm commented:
hi @chavafg , thanks for the review. I resolved many of your comments in this new version. I forgot one change though, gonna send another version. :)

v1 -> v2:
 * `integration/openshift/e2e/gen_test_cases.sh`:
   * Properly quoted variables
 * `integration/openshift/lib.sh`
   * small indentation fix

wainersm commented:
v2 -> v3:

 * `integration/openshift/e2e/runner.sh`
   * Replaced `exit 1` with `die`

cpmeadors reviewed (changes requested):
I tried again to run the tests and again the check for the admission controller was failing.  I looked at the logs for the admission controller:

2020/09/28 14:46:18 http: TLS handshake error from 10.129.0.1:45802: remote error: tls: bad certificate

That would explain why the pod never gets runtimeClass set.  I installed the controller by running run_tests.sh with CI=true.

jodh-intel commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,104 @@
+
+Usage $0: [-c] [-d] [-r NAME], where:
+  -h, --help   Print this message.
+  -c           Prepare the cluster to run the tests.
```
Might be worth adding the word "configure" to the text as a mnemonic for those reading this text.

wainersm replied:
Sure. On next version.

jodh-intel commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,104 @@
+EOF
+}
+
+[ -n "$KUBECONFIG" ] || die "The KUBECONFIG variabe is unset"
```
s/variabe/variable/

wainersm replied:
Annotated to next version.

jodh-intel commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,104 @@
+  -h, --help   Print this message.
+  -c           Prepare the cluster to run the tests.
+  -d           Run the tests in dry-run mode.
+  -r           The runtimeClass name (default is '$kata_runtimeclass'). This
+               option should not be used with -c.
```
I don't see a check in the code for this?

wainersm replied:
Good idea, documentation is rarely enough :)
Going to add a check.

jodh-intel commented on integration/openshift/e2e/test_cases.txt:
```diff
@@ -0,0 +1,2367 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Auto-generated by gen_test_cases.sh (2020-09-18)
```
Why is this (relatively huge) config file being stored if it can be generated by a script included on this PR?

wainersm replied:
Good question, I did this way but I confess I'm not sure about it.

In my opinion not passing a list of tests explicitly is error prone. For this case of the OpenShift tests, there isn't any guarantee that the list will be the same between two executions because the `openshift-tests` binary might change (by mistake) so the tests list. Therefore, having a fixed list of tests ensure we are comparing the same set of tests between executions (at least nominally, because the test implementation might change after all).

The drawbacks are that it should to be re-generated every time we bump the version (see #github.com/kata-containers/runtime/pull/2969) of the test binary or when a new filter is needed; and also the file size.

About the file size, currently it simply includes all the tests from OpenShift e2e but I would like to work on narrowing the list down to just the ones which makes sense for exercising Kata Containers.

What's your opinion @jodh-intel ?

jodh-intel commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,156 @@
+# ATTENTION: the caller of this lib should have the tests_dir exported
+#            correctly
+if [ ! -d "${tests_dir}" ]; then
+	echo "Unable to find the Kata Container tests directory."
```
Nit: `echo >&2 "ERROR: Unable ...`.

wainersm replied:
Also should use Kata ContainerS in plural.

jodh-intel commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,156 @@
+# Contain helper functions and variables for the OpenShift integration
+# test scripts.
+#
+
```
All the other scripts on this PR use `set -e`, so should this one too?

jodh-intel commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,156 @@
+	info "Add the ${kata_runtimeclass} runtimeClass"
+	if is_runtimeclass_present; then
+		info "runtimeClass is already present"
+	else
```
Might be worth adding a `return` in the `if` block to avoid the extra indent for `else`/`fi` which always looks odd when combined with a heredoc ;)

jodh-intel reviewed (commented):
Thanks @wainersm. A few comments.

wainersm commented:
> I tried again to run the tests and again the check for the admission controller was failing. I looked at the logs for the admission controller:
> 
> 2020/09/28 14:46:18 http: TLS handshake error from 10.129.0.1:45802: remote error: tls: bad certificate
> 
> That would explain why the pod never gets runtimeClass set. I installed the controller by running run_tests.sh with CI=true.

Yes, something is wrong on the admission controller so that it doesn't work in your cluster. Yesterday I installed a fresh 4.5.7 cluster and it worked out.

That might be some security feature active in your cluster which prevents the webhook pod from properly accessing the secret key?

wainersm commented:
v3 -> v4:
 * `integration/openshift/e2e/runner.sh`
   * Added checking to ensure the -c and -r options aren't set at same time
   * Fixed spelling
   * Fixed quoting around variable
 * `integration/openshift/lib.sh`
   * Use `set -e`
   * Cosmetic changes suggested by @jodh-intel

cpmeadors reviewed (approved):
I can't fully run the code due to an issue with the admission controller, https://bugzilla.redhat.com/show_bug.cgi?id=1884305, but the code up to that point worked.  The setup and env checks worked.  The code is clear and well documented.  File more bugs when admission controller issue is resolved.

fidencio commented:
/test

fidencio commented:
Wainer, could you check whether the CI errors are the same I'm facing here https://github.com/kata-containers/tests/pull/2921 ?

jodh-intel commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,107 @@
+
+while getopts "hcdr:" opt; do
+	case ${opt} in
+		c)
+			configure_cluster="true"
+			;;
```
Entirely up to you, but since all these cases are very brief, you could collapse them down into single lines maybe?

```bash
c) configure_cluster="true" ;;
d) dry_run="true" ;;
    :
```
```

jodh-intel commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,107 @@
+
+# Save the tests results in junit format.
+#
+output_dir="${script_dir}/../../../_output/tests/openshift-e2e"
```
Why all the path traversal? Could you do `output_dir=$(mktemp -d)` instead?

wainersm replied:
I'm thinking on OpenShift CI when I set the output directory to `${tests_repo}/_output/tests/openshift-e2e`, because this way the generated junit file can be indexed properly. If I use a temp dir then the path is unpredictable to the caller CI scripts.

jodh-intel replied:
Hmm. Maybe the caller could pass an "output direction parameter" to avoid this script having to make assumptions and use paths like this (which look confusing and error prone potentially to my eyes ;)

wainersm replied:
I think you're right @jodh-intel, it deserves a parameter for the output in the `runner.sh`

Let me take that opportunity to ask one thing related... at begin I was attempted to use `KATA_TESTS_LOGDIR` (exported in `tests/lib/common.bash`) but I got the impression test scripts don't use it. Then the other day I saw in one of your PRs it is being used but only when `CI=true`. So I am curious about what is the orientation for the use of that variable (and others KATA_TESTS_*) in the scripts.

jodh-intel replied:
That variable can be used by tests if needed. I think I added it for https://github.com/kata-containers/tests/blob/master/tracing/tracing-test.sh, but no other test really needs to use the feature since most just log to `stdout`/`stderr` and/or write to the journal.

Note that Jenkins collects up all these artifacts by calling the following at the end of a test run:

- https://github.com/kata-containers/tests/blob/master/.ci/teardown.sh

jodh-intel commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,107 @@
+
+cmd+=" -f -"
+info "Run the tests"
+{ set -x; cat $test_cases_file | grep -v ^# | ${cmd}; } || die "FAILED"
```
Nit: Quote the grep pattern as GitHub thinks it's a comment :)

wainersm replied:
Even `vim` cannot highlight it properly :D

jodh-intel replied:
Ouch. It does if you backslash the hash though ;)

jodh-intel commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,157 @@
+			$(find -type f -executable -name "openshift-tests"))"
+	popd
+	} &>/dev/null
+	[ $? -eq 0 ] || return 1
```
Is this correct? Return error and no string if the build is successful? But actually it isn't the build but the `popd` return value which is almost guaranteed to succeed isn't it (since we've already done a successful `pushd`)?

wainersm replied:
The idea is to return the string or error if the build failed. But you are right, it is going to get the success of `popd`. I was almost sure I had tested this function...anyway, I will fix it by checking if `$cmd` is non-empty (or `return 1`).

jodh-intel commented on integration/openshift/lib.sh:
```diff
@@ -0,0 +1,157 @@
+
+	# Create a hello world pod and check if it is using the runtimeClass.
+	#
+	curl -s https://raw.githubusercontent.com/openshift/origin/master/examples/hello-openshift/hello-pod.json |\
```
Might be worth add a `-L` here.

wainersm replied:
BTW, @fidencio told me the other day that `apply -f url` can handle this, so I will remove this `curl` call.

jodh-intel reviewed (commented):
Thanks @wainersm.

c3d commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,77 @@
+
+if [ ${#skip_patterns[@]} -gt 0 ]; then
+	cat $tmp_tests_file | \
+		grep -v -f ${skip_file} >> $tmp_tests_file
```
Any reason to use use extended regexps in one case and not the other?

wainersm replied:
Yes, the reason is: I did a mistake. :) Gonna fix that.

c3d replied:
> Yes, the reason is: I did a mistake. :) Gonna fix that.

That sounds like a good reason ;-)

c3d commented on integration/openshift/e2e/run_tests.sh:
```diff
@@ -0,0 +1,28 @@
+
+info "Run OpenShift e2e tests"
+
+if [ "$OPENSHIFT_TESTS_DRY_RUN" == "true" ];then
```
Who sets `OPENSHIFT_TEST_DRY_RUN`?

(asking because if some poor user sets it, they may be surprised if setting it to `1` does not work)

Also a bit confused by having both `OPENSHIFT_TEST_DRY_RUN` and `dry_run` in the scripts.

jodh-intel replied:
Agreed. If we really need both, I'd do something like the following as early in the script as possible:

```bash
[ "$OPENSHIFT_TEST_DRY_RUN" = true ] && dry_run="true" and then only refer to `$dry_run` if possible.
```

wainersm replied:
> Who sets `OPENSHIFT_TEST_DRY_RUN`?

Oh, I forgot to initialize the variable to `false` if not set. The caller of `make openshift-e2e` can export this variable to get the tests running on dry-run mode.

ps: The `CI` variable should be initialized likewise, and I will fix that.

> 
> (asking because if some poor user sets it, they may be surprised if setting it to `1` does not work)
> 
> Also a bit confused by having both `OPENSHIFT_TEST_DRY_RUN` and `dry_run` in the scripts.

Ok, the `runner.sh` is the "backend" script which just run the tests. While developing that script I found handy to have this `dry_run` variable set (which turn on the dry-run mode), because I didn't always needed to run all the tests for some unrelated changes.

I thought that users in general could be interested on this dry-run feature as well. But now that you asked, maybe I should simply remove it...if someone finds value on it then later I can put it back. What do you think @c3d @jodh-intel ?

c3d replied:
No objection to removing it, no objection to leaving it either.

c3d commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,107 @@
+This script runs a sub-set of the OpenShift e2e tests in other to test Kata
+Containers in a running OpenShift cluster.
+
+It assumes that the KUBECONFIG variable is exported and it points to a cluster
```
nit: "and points" or "and that it points"

c3d commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,107 @@
+    to use that runtimeClass.
+
+If you choose to have the cluster configured by this script, beware that it
+will not return to its original state on the end.
```
This is a bit frightening ;-)

wainersm replied:
Cluster's life is hard my friend :)

c3d commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,107 @@
+# admission controller is active.
+#
+if [ "$configure_cluster" == "true" ]; then
+	[ "$default_kata_runtimeclass" == "$kata_runtimeclass" ] || \
```
What is the point of that test? I seem to remember a recent discussion regarding alternate runtime classes, e.g. `kata-qemu-virtiofs-xattr-2.0-softmmu-relaxed-nowarnings`.

wainersm replied:
I made this OpenShift e2e tests support thinking in two use cases:

1. Manual testing. The person should know what she/he is doing, is likely the cluster is already configured but, to be sure, the `runner.sh` is going to do some checking. The user might have configured a `runtimeClass` with, e.g., `my-kata-runtime` and this is going to break the checking. So It was introduced the `-r` option, allowing to overwrite the default `runtimeClass` name used on the check.
1. Automated testing. In this case "the thing" (likely an CI) doesn't have the cluster full configured, and here comes the `-c` option to rescue. It is used the http://github.com/kata-containers/tests/tree/master/kata-webhook code to configure the admission controller, which has hard-coded the `runtimeClass` name to `kata`.

Therefore, if someone passes `-c` and `-r` options together then the configuration step is going to use the default `runtimeClass` name (i.e. `kata`)  and the checking is going to use whatever was passed by the user, and will fail. That's the reason why I added that test.

Makes sense?

c3d replied:
It does. I would expand the error message then:

```
die "Configuring the cluster uses kata-webhook, which deploys only the default runtimeClass $default_kata_runtime_class, so unable to test with $runtime_class"
```

It's a bit verbose, but at least you know where to look.

wainersm replied:
cool, I will improve the message. Thanks @c3d

wainersm commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,77 @@
+# Edit the filter_patterns, exclude_patterns, and skip_patterns variables
+# and re-run this script in order to change the tests selection.
+#
+set -e
```
Yeah, I set this option and found a bug. So absolutely worth leaving it. Thanks!

wainersm commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,77 @@
+fi
+
+print_header > $test_cases_file
+cat $tmp_tests_file | sort -u >> $test_cases_file
```
yes, going to include a `cleanup()` function.

wainersm commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,77 @@
+source "$script_dir/../lib.sh"
+
+# Tests that should be included. Leave it empty to pick them all.
+filter_patterns=(
```
Hi @c3d , hey, thanks for reviewing too!

> I would not call them patterns.

How would you call them? I'm really expecting grep-like (extended) patterns in quotes as you mentioned.

> 
> The problem is that there are two expansions at play here, the first one being the shell, the second one from your `grep` below. So if you have `*.znort` here, this will really be expanded by the shell first.
> 

I didn't understand the second expansion you mentioned. Because It writes the patterns to a file, then use the `grep -f` to read them from the same file.
 
> Maybe having a commended-out example of the kind of pattern you expect here? I suspect it would be grep-like patterns in quotes, i.e. `'foo*'` would not match `fooa` (not glob) but would match `foooo` (because `grep`).

Yeah, it deserves a better text explaining that.

c3d replied:
> Hi @c3d , hey, thanks for reviewing too!
> 
> > I would not call them patterns.
> 
> How would you call them? I'm really expecting grep-like (extended) patterns in quotes as you mentioned.

Sorry, should have been clear. I would call them "regexps", that way you know it's not glob patterns.

> 
> > The problem is that there are two expansions at play here, the first one being the shell, the second one from your `grep` below. So if you have `*.znort` here, this will really be expanded by the shell first.
> 
> I didn't understand the second expansion you mentioned. Because It writes the patterns to a file, then use the `grep -f` to read them from the same file.

If in your script someone writes 

```
filter_patterns=(
    *.txt
)
```

then the shell first interprets that as a glob pattern, so you end up with an array containing maybe something like `a.txt`, `b.txt`, and so on.


> 
> > Maybe having a commended-out example of the kind of pattern you expect here? I suspect it would be grep-like patterns in quotes, i.e. `'foo*'` would not match `fooa` (not glob) but would match `foooo` (because `grep`).
> 
> Yeah, it deserves a better text explaining that.

jodh-intel commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,104 @@
+	rm -f "${inclusion_file}" "${exclusion_file}" "${skip_file}"
+}
+
+trap cleanup EXIT
```
Nit: Move this call to `setup()`.

jodh-intel commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,104 @@
+
+function print_header()
+{
+	cat <<EOF
```
Nit: You could use `<<-EOF` and then tab indent the here doc lines to make the function look neater ;)

jodh-intel commented on integration/openshift/e2e/gen_test_cases.sh:
```diff
@@ -0,0 +1,104 @@
+
+function cleanup()
+{
+	rm -f "${inclusion_file}" "${exclusion_file}" "${skip_file}"
```
Nit: What about `$tmp_tests_file`? You could use a slightly different idiom to make cleanup more reliable:

```bash
files_to_delete=()

    :

files_to_delete+=("/tmp/foo")
files_to_delete+=("/tmp/bar")
    :

[ "${#files_to_delete[@]}" -gt 0 ] && rm -f "${files_to_delete[@]}"
```

wainersm replied:
done.

jodh-intel commented on integration/openshift/e2e/runner.sh:
```diff
@@ -0,0 +1,119 @@
+	#
+	if [ "$configure_cluster" == "true" ]; then
+		[ "$default_kata_runtimeclass" == "$kata_runtimeclass" ] || \
+			die "To configure the cluster it should use the default runtimeClass $default_kata_runtime_class, so do not use the -c and -r options together"
```
Nit: super long line ;)

wainersm replied:
I tried to improve, let me know what you think.

jodh-intel reviewed (approved):
Thanks @wainersm.

lgtm

jodh-intel commented:
/test-ubuntu

c3d commented:
Remaining issues have been addressed. Merging.
